lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1795,7 +1810,7 @@ class MongoClientInterface {
                     };
                     log.error(
                         'internalListObjectV1: error listing objects', logObj);
-                    baseStream.destroy();
+                    cleanupStreams();
```

Comment:
it seems we have already a mecanism to handle the cleanup (with a bug indeed):
```
                resolvePhdKey.on('end', () => {
                    baseStream.emit('end');
                });
```

instead of adding a "global" cleanup called multiple times, and difficult to anticipate where it will be called and if it will always be called exactly once: it seems better to let each transform handle its own cleanup AND propagate the signals/errors.

Reply:
I fear "just" fixing the current code will likely let the issue come back someday if we miss another cleanup, which is the mistake we did last time. The current leak was already there, but since we did not see it, we forgot to add cleanup for this case. Given all the code paths here, I am definitely not confident that fixing only the current leak (line 1853) will fix everything for all code paths, unless we clean everything in a consistent way.

I vote for a wrapper at the en (cbOnce), so we clean every time we complete one function call. However I will need to spend time again to confirm this still works, due to the mix of recursive calls, events, and callbacks...

Reply:
Is it what you had in mind? https://github.com/scality/Arsenal/pull/2264/commits/d2d5391bdcac821aad1d91f6240993ba3cd0af5c

Reply:
still some "duplicated" `stream.destroy`:

```
        skip.setListingEndCb(() => {
            stream.emit('end');
            stream.destroy();
        });
        skip.setSkipRangeCb(range => {
            // stop listing this key range
            stream.destroy();
```

Reply:
I'm not sure we should remove this, I need to test first, sending "end" seem more appropriate, to keep the cleanup in a single place

Reply:
Note: This PR will not be merged till we re-confirm the latest changes are stable on a real platform. Expecting no merge before next monday.

Reply:
So we can remove the first destroy indeed (but keep the end event). For the second, we indeed must keep the cleanup, otherwise we may get more data from the stream in the parent function, making the whole thing never returning the callback.

- I also updated the logic to only destroy the baseStream if it is not the `stream` object.
- All streams are set to `null` -> not formally needed, but you suggested it and it makes it explicit for the garbage collector that we don't use the variable anymore, so likely to be cleaned faster than before.

https://github.com/scality/Arsenal/pull/2264/commits/8961aa53c1a9b9ee637a5f12376e7ee15550299e